relayer/checkpoint.go

```diff
@@ -0,0 +1,212 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package relayer
+
+import (
+	"container/heap"
+	"strconv"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/awm-relayer/database"
+	"github.com/ava-labs/awm-relayer/utils"
+	"go.uber.org/zap"
+)
+
+//
+// checkpointManager commits keys to be written to the database in a thread safe manner.
+//
+
+type checkpointManager struct {
+	logger                   logging.Logger
+	database                 database.RelayerDatabase
+	writeSignal              chan struct{}
+	relayerID                database.RelayerID
+	queuedHeightsAndMessages map[uint64]*messageCounter
+	committedHeight          uint64
+	lock                     *sync.RWMutex
+	pendingCommits           *utils.UInt64Heap
+	finished                 chan uint64
+}
+
+func newCheckpointManager(logger logging.Logger, database database.RelayerDatabase, writeSignal chan struct{}, relayerID database.RelayerID) *checkpointManager {
+	h := &utils.UInt64Heap{}
+	heap.Init(h)
+	return &checkpointManager{
+		logger:                   logger,
+		database:                 database,
+		writeSignal:              writeSignal,
+		relayerID:                relayerID,
+		queuedHeightsAndMessages: make(map[uint64]*messageCounter),
+		lock:                     &sync.RWMutex{},
```

Comment:
So I think we need to initialize the checkpointManager with the correct starting block height as `committedHeight`. Right now we have a race condition where the first call to `stageCommittedHeight()` will set `committedHeight` to that height. So if we process a block from the chain head before any of the catchup blocks, we will write the chain head to the database accidentally

Reply:
Great call out. I've refactored the checkpoint manager constructor to initialize `committedHeight`. We fetch this for each application relayer using the `CalculateStartingBlockHeight`, which I've moved to `database/utils.go`. The `Listener` begins listening from the minimum starting block height of all its composite application relayers. The checkpoint manager will only prepare heights that are greater than its `committedHeight` to account for possible re-processing of already committed blocks on startup.